### PR TITLE
UCP/WIREUP: replay pending requests via common tmp queue

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What
Replay pending requests via common tmp queue

## Why ?
Avoid fragmented proto_v2 request reset multiple times on multi
lane endpoints during replay in the end of wireup on each wireup
endpoint

### CODE REVIEW

File: src/ucp/rma/flush.c [hunk 1]
```diff
@@ -201,14 +221,32 @@ ucs_status_t ucp_ep_flush_progress_pending(uct_pending_req_t *self)
...
+    ucs_assert(!(req->send.flush.started_lanes & UCS_BIT(lane)));
+
     status = uct_ep_flush(ucp_ep_get_lane(ep, lane), req->send.flush.uct_flags,
                           &req->send.state.uct_comp);
-    ucs_trace("flushing ep %p lane[%d]=%p: %s", ep, lane,
-              ucp_ep_get_lane(ep, lane), ucs_status_string(status));
+    ucs_trace_req("req %p: flush ep %p lane[%d]=%p: %s", req, ep, lane,
+                  ucp_ep_get_lane(ep, lane), ucs_status_string(status));
     if (status == UCS_OK) {
+        req->send.flush.started_lanes |= UCS_BIT(lane);
```
Comment (yosefe) [inline src/ucp/rma/flush.c:243]: ohhh...
maybe add wrapper for --comp.count + updating started lanes and reuse it in lines 115-120?

File: src/ucp/rma/flush.c [hunk 2]
```diff
@@ -273,8 +311,7 @@ void ucp_ep_flush_request_ff(ucp_request_t *req, ucs_status_t status)
      * this request from its pending queue
      */
     int num_comps = req->send.flush.num_lanes -
-                    ucs_popcount(req->send.flush.started_lanes)
-                    + 1;
```
Comment (yosefe) [inline src/ucp/rma/flush.c:277]: why?

File: src/ucp/core/ucp_ep.c [hunk 3]
```diff
@@ -443,6 +444,11 @@ void ucp_ep_destroy_base(ucp_ep_h ep)
     ucp_ep_release_id(ep);
     ucs_list_del(&ep->ext->ep_list);
 
+    if (ep->ext->progress_id != UCS_CALLBACKQ_ID_NULL) {
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:447]: AFAIK it's already checked in uct_worker_progress_unregister_safe

File: src/ucp/rma/flush.c [hunk 4]
```diff
@@ -99,32 +113,33 @@ static void ucp_ep_flush_progress(ucp_request_t *req)
                       ucs_status_string(status));
         if (status == UCS_OK) {
             req->send.flush.started_lanes |= UCS_BIT(lane);
+            ucs_assert(req->send.state.uct_comp.count > 0);
```
Comment (yosefe) [inline src/ucp/rma/flush.c:116]: can we unite checking count>0 and decrementing it to a helper function?
probably could be reused in other places too

File: src/ucp/rma/flush.c [hunk 5]
```diff
@@ -201,14 +221,32 @@ ucs_status_t ucp_ep_flush_progress_pending(uct_pending_req_t *self)
 
     ucs_assert(!(req->flags & UCP_REQUEST_FLAG_COMPLETED));
 
+    if (ep->flags & UCP_EP_FLAG_RESTART_FLUSH) {
+        uct_worker_progress_register_safe(ep->worker->uct,
+                                          ucp_ep_flush_resume_slow_path_callback,
+                                          req, 0, &req->send.flush.prog_id);
+        /* Request was detached from pending and should be scheduled again */
+        ucs_assert((req->send.lane == 0));
```
Comment (yosefe) [inline src/ucp/rma/flush.c:229]: extra ()
and use assertv

File: src/ucp/rma/flush.c [hunk 6]
```diff
@@ -201,14 +221,32 @@ ucs_status_t ucp_ep_flush_progress_pending(uct_pending_req_t *self)
...
+        ucs_assert(!req->send.flush.sw_started);
```
Comment (yosefe) [inline src/ucp/rma/flush.c:230]: ucs_assertv to print req ptr

File: src/ucp/rma/flush.c [hunk 7]
```diff
@@ -83,6 +96,7 @@ static void ucp_ep_flush_progress(ucp_request_t *req)
         uct_ep = ucp_ep_get_lane(ep, lane);
         if (uct_ep == NULL) {
             req->send.flush.started_lanes |= UCS_BIT(lane);
+            ucs_assert(req->send.state.uct_comp.count > 0);
```
Comment (yosefe) [inline src/ucp/rma/flush.c:99]: ucs_assertv to print req ptr

File: src/ucp/rma/flush.c [hunk 8]
```diff
@@ -61,20 +69,25 @@ static void ucp_ep_flush_progress(ucp_request_t *req)
...
-                      destroyed_lanes, ucs_popcount(destroyed_lanes));
             ucs_assert(!(req->send.flush.started_lanes & destroyed_lanes));
+            ucp_trace_req(req,
+                          "ep %p flush lanes changed from %d to %d, "
+                          "destroyed_lanes 0x%x, reducing completion "
+                          "count by %d",
+                          ep, req->send.flush.num_lanes, num_lanes,
+                          destroyed_lanes, ucs_popcount(destroyed_lanes));
             req->send.state.uct_comp.count -= ucs_popcount(destroyed_lanes);
+            ucs_assert(req->send.state.uct_comp.count >= 0);
```
Comment (yosefe) [inline src/ucp/rma/flush.c:80]: ucs_assertv to print req ptr

File: src/ucp/rma/flush.c [hunk 9]
```diff
@@ -45,14 +50,17 @@ static void ucp_ep_flush_progress(ucp_request_t *req)
     int diff;
     ucp_lane_map_t destroyed_lanes;
 
+    ucs_assert(!(ep->flags & UCP_EP_FLAG_RESTART_FLUSH));
```
Comment (yosefe) [inline src/ucp/rma/flush.c:53]: ucs_assertv to print ep ptr

File: src/ucp/rma/flush.c (hunk 2 above)
Comment (evgeny-leksikov) [reply src/ucp/rma/flush.c:277]: IIUC 1 was added for pending started lane, which is not counted now

File: src/ucp/wireup/wireup.c [hunk 10]
```diff
@@ -419,15 +419,122 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+
+void ucp_wireup_remote_connect_lanes(ucp_ep_h ucp_ep, int ready)
+{
+    ucp_lane_index_t lane;
+    ucp_wireup_ep_t *wireup_ep;
+    uct_ep_h uct_ep;
+
+    for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
+        uct_ep = ucp_ep_get_lane(ucp_ep, lane);
+        if (ucp_wireup_ep_test(uct_ep)) {
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:515]: if (!...) continue

File: src/ucp/wireup/wireup.c [hunk 11]
```diff
@@ -419,15 +419,122 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+            wireup_ep = ucp_wireup_ep(uct_ep);
+
+            ucs_assert(wireup_ep != NULL);
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:518]: redundant

File: src/ucp/wireup/wireup.c [hunk 12]
```diff
@@ -874,6 +981,13 @@ ucp_wireup_replay_pending_request(uct_pending_req_t *self, ucp_ep_h ucp_ep)
 
     ucs_assert(req->send.ep == ucp_ep);
 
+    if (req->flags & UCP_REQUEST_FLAG_PROTO_SEND) {
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:984]: maybe print it in the next 'if'?

File: src/ucp/wireup/wireup.c [hunk 13]
```diff
@@ -887,10 +1001,16 @@ void ucp_wireup_replay_pending_requests(ucp_ep_h ucp_ep,
 {
     uct_pending_req_t *uct_req;
 
+    if (ucs_queue_is_empty(tmp_pending_queue)) {
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:1004]: seems redundant

File: src/ucp/core/ucp_ep.h [hunk 14]
```diff
@@ -463,6 +464,7 @@ typedef struct ucp_ep_ext {
     ucs_ptr_map_key_t             local_ep_id;   /* Local EP ID */
     ucs_ptr_map_key_t             remote_ep_id;  /* Remote EP ID */
     ucp_err_handler_cb_t          err_cb;        /* Error handler */
+    uct_worker_cb_id_t            progress_id;   /* ID of wireup progress function */
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.h:467]: can we avoid adding this to ep-ext?

File: src/ucp/wireup/wireup.c [hunk 15]
```diff
@@ -419,15 +419,122 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+    for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
+        uct_ep = ucp_ep_get_lane(ucp_ep, lane);
+        if (ucp_wireup_ep_test(uct_ep)) {
+            wireup_ep = ucp_wireup_ep(uct_ep);
+
+            ucs_assert(wireup_ep != NULL);
+            ucs_assert(wireup_ep->super.uct_ep != NULL);
+            ucs_assert(wireup_ep->flags & UCP_WIREUP_EP_FLAG_LOCAL_CONNECTED);
+
+            ucp_ep            = wireup_ep->super.ucp_ep;
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:522]: why overwrite?

File: src/ucp/rma/flush.c (hunk 1 above)
Comment (evgeny-leksikov) [reply src/ucp/rma/flush.c:243]: yeah, let's see how CI goes... ep flush looks like corner circle or inf polygon :) then clean up the code

File: src/ucp/core/ucp_ep.h (hunk 14 above)
Comment (yosefe) [reply src/ucp/core/ucp_ep.h:467]: better oneshot - to not increase size of ep_ext

File: src/ucp/core/ucp_ep.h [hunk 16]
```diff
@@ -101,6 +101,7 @@ enum {
     UCP_EP_FLAG_STREAM_HAS_DATA        = UCS_BIT(5), /* EP has data in the ext.stream.match_q */
     UCP_EP_FLAG_ON_MATCH_CTX           = UCS_BIT(6), /* EP is on match queue */
     UCP_EP_FLAG_REMOTE_ID              = UCS_BIT(7), /* remote ID is valid */
+    UCP_EP_FLAG_RESTART_FLUSH          = UCS_BIT(8), /* Flush ops have to be restarted from pending queue */
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_ep.h:104]: rename to `UCP_EP_FLAG_BLOCK_FLUSH`

**Comment by yosefe** [comment]
@dmitrygx @hoopoepg can you pls review?

File: src/ucp/core/ucp_ep.c [hunk 17]
```diff
@@ -413,12 +413,19 @@ static int ucp_ep_set_failed_remove_filter(const ucs_callbackq_elem_t *elem,
     return 0;
 }
 
+static int ucp_ep_wireup_eps_progress_filter(const ucs_callbackq_elem_t *elem,
+                                      void *arg)
```
Comment (hoopoepg) [inline src/ucp/core/ucp_ep.c:417]: strange alignment

File: src/ucp/wireup/wireup.c [hunk 18]
```diff
@@ -419,15 +419,128 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+}
+
+/*
+ * We switch the endpoints in this function (instead in wireup code) since
+ * this is guaranteed to run from the main thread.
+ */
+unsigned ucp_wireup_eps_progress(void *arg)
+{
+    ucp_ep_h ucp_ep = arg;
+    ucp_wireup_ep_t *wireup_ep = NULL;
```
Comment (hoopoepg) [inline src/ucp/wireup/wireup.c:460]: alignment

**Review by hoopoepg — APPROVED** [review]

File: src/ucp/rma/flush.c [hunk 19]
```diff
@@ -15,17 +15,43 @@
...
+                "req %p: started_lanes 0x%x, new_started_lanes 0x%x", req,
+                req->send.flush.started_lanes, new_started_lanes);
+
+    req->send.state.uct_comp.count += diff;
+    req->send.flush.started_lanes  |= new_started_lanes;
+
+    ucp_trace_req(req,
+                  "ep %p flush changed completion count %d by %d, "
+                  "started lanes 0x%x by 0x%x", req->send.ep,
+                  req->send.state.uct_comp.count, diff,
```
Comment (dmitrygx) [inline src/ucp/rma/flush.c:36]: but `req->send.state.uct_comp.count` contains a result after applying `diff`. maybe rephrase it:
```
"ep %p flush changed completion count by %d - result %d, "
```

File: src/ucp/rma/flush.c [hunk 20]
```diff
@@ -45,45 +71,49 @@ static void ucp_ep_flush_progress(ucp_request_t *req)
...
-                      ep, req, req->send.flush.num_lanes, num_lanes,
-                      destroyed_lanes, ucs_popcount(destroyed_lanes));
             ucs_assert(!(req->send.flush.started_lanes & destroyed_lanes));
-            req->send.state.uct_comp.count -= ucs_popcount(destroyed_lanes);
+            ucp_trace_req(req,
+                          "ep %p flush lanes changed from %d to %d, "
+                          "destroyed_lanes 0x%x",
+                          ep, req->send.flush.num_lanes, num_lanes,
+                          destroyed_lanes);
+            ucp_ep_flush_request_change_uct_cmp(req, -ucs_popcount(destroyed_lanes), 0);
```
Comment (dmitrygx) [inline src/ucp/rma/flush.c:98]: does it fit 80 symbols per line?

File: src/ucp/rma/flush.c [hunk 21]
```diff
@@ -15,17 +15,43 @@
...
-static void ucp_ep_flush_error(ucp_request_t *req, ucs_status_t status)
+static void
+ucp_ep_flush_request_change_uct_cmp(ucp_request_t *req, int diff,
+                                    ucp_lane_map_t new_started_lanes)
+{
+    ucs_assertv((req->send.state.uct_comp.count + diff) >= 0,
+                "req %p: comp %p count %d, diff %d", req,
+                &req->send.state.uct_comp, req->send.state.uct_comp.count,
+                diff);
+    ucs_assertv((req->send.flush.started_lanes & new_started_lanes) == 0,
```
Comment (dmitrygx) [inline src/ucp/rma/flush.c:26]: maybe
```suggestion
    ucs_assertv(!(req->send.flush.started_lanes & new_started_lanes),
```
also, apply for other places, because it both uses `!(expr)` and `(expr) == 0`. but should be aligned

File: src/ucp/wireup/wireup.c [hunk 22]
```diff
@@ -419,15 +419,128 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+    ucp_wireup_ep_t *wireup_ep;
+    uct_ep_h uct_ep;
+
+    for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
+        uct_ep = ucp_ep_get_lane(ucp_ep, lane);
+        if (!ucp_wireup_ep_test(uct_ep)) {
+            continue;
+        }
+
+        wireup_ep = ucp_wireup_ep(uct_ep);
```
Comment (dmitrygx) [inline src/ucp/wireup/wireup.c:527]:
```suggestion
        uct_ep    = ucp_ep_get_lane(ucp_ep, lane);
        wireup_ep = ucp_wireup_ep(uct_ep);
        if (wireup_ep == NULL) {
            continue;
        }
```

File: src/ucp/wireup/wireup.c [hunk 23]
```diff
@@ -419,15 +419,128 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+        uct_ep = ucp_ep_get_lane(ucp_ep, lane);
+        if (ucp_wireup_ep_test(uct_ep)) {
+            wireup_ep = ucp_wireup_ep(uct_ep);
+            ucs_assert(wireup_ep->flags & UCP_WIREUP_EP_FLAG_READY);
+            ucs_assert(wireup_ep->super.uct_ep != NULL);
+            if (wireup_ep->pending_count != 0) {
+                return 1;
+            }
         }
     }
```
Comment (dmitrygx) [inline src/ucp/wireup/wireup.c:438]:
```suggestion
    for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
        uct_ep    = ucp_ep_get_lane(ucp_ep, lane);
        wireup_ep = ucp_wireup_ep(uct_ep);
        if (wireup_ep == NULL) {
            continue;
        }

        ucs_assert(wireup_ep->flags & UCP_WIREUP_EP_FLAG_READY);
        ucs_assert(wireup_ep->super.uct_ep != NULL);
        if (wireup_ep->pending_count != 0) {
            return 1;
        }
    }
```

File: src/ucp/wireup/wireup.c [hunk 24]
```diff
@@ -419,15 +419,128 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+            wireup_ep = ucp_wireup_ep(uct_ep);
+
+            ucs_queue_for_each_extract(uct_req, &wireup_ep->pending_q, priv, 1) {
+                ucs_queue_push(&tmp_pending_queue, ucp_wireup_ep_req_priv(uct_req));
+                ucp_worker_flush_ops_count_dec(ucp_ep->worker);
+            }
+
+            /* Switch to real transport and destroy proxy endpoint (aux_ep as well) */
+            ucp_proxy_ep_replace(&wireup_ep->super);
+        }
```
Comment (dmitrygx) [inline src/ucp/wireup/wireup.c:502]:
```suggestion
        uct_ep    = ucp_ep_get_lane(ucp_ep, lane);
        wireup_ep = ucp_wireup_ep(uct_ep);
        if (wireup_ep  == NULL) {
            continue;
        }

        ucs_queue_for_each_extract(uct_req, &wireup_ep->pending_q, priv, 1) {
            ucs_queue_push(&tmp_pending_queue, ucp_wireup_ep_req_priv(uct_req));
            ucp_worker_flush_ops_count_dec(ucp_ep->worker);
        }

        /* Switch to real transport and destroy proxy endpoint (aux_ep as well) */
        ucp_proxy_ep_replace(&wireup_ep->super);
```

**Review by dmitrygx — APPROVED** [review]

File: src/ucp/rma/flush.c [hunk 25]
```diff
@@ -185,12 +214,41 @@ static unsigned ucp_ep_flush_resume_slow_path_callback(void *arg)
...
     ucp_flush_check_completion(req);
     return 0;
 }
 
+static void ucp_ep_flush_slow_path_resched(ucp_ep_h ep, ucp_request_t *req)
+{
+    if (ep->flags & UCP_EP_FLAG_BLOCK_FLUSH) {
+        /* Request was detached from pending and should be scheduled again */
+        ucs_assertv(!(UCS_BIT(req->send.lane) & req->send.flush.started_lanes),
+                    "req %p: lane %d, started_lanes 0x%x", req,
```
Comment (yosefe) [inline src/ucp/rma/flush.c:231]: "req=%p lane=%d started_lanes=0x%x"

File: src/ucp/rma/flush.c [hunk 26]
```diff
@@ -185,12 +214,41 @@ static unsigned ucp_ep_flush_resume_slow_path_callback(void *arg)
...
+                    req->send.lane, req->send.flush.started_lanes);
+        /* Only lanes connected to iface can be started/flushed before
+         * wireup is done */
```
Comment (yosefe) [inline src/ucp/rma/flush.c:234]: elaborate why: because connect2iface does not create wireup_ep without cm mode

File: src/ucp/rma/flush.c [hunk 27]
```diff
@@ -201,14 +259,24 @@ ucs_status_t ucp_ep_flush_progress_pending(uct_pending_req_t *self)
 
     ucs_assert(!(req->flags & UCP_REQUEST_FLAG_COMPLETED));
 
+    if (ep->flags & UCP_EP_FLAG_BLOCK_FLUSH) {
+        ucp_ep_flush_slow_path_resched(ep, req);
```
Comment (yosefe) [inline src/ucp/rma/flush.c:263]: ucp_ep_flush_request_resched

File: src/ucp/rma/flush.c [hunk 28]
```diff
@@ -201,14 +259,24 @@ ucs_status_t ucp_ep_flush_progress_pending(uct_pending_req_t *self)
...
+        return UCS_OK;
+    }
+
+    ucs_assert(lane != UCP_NULL_LANE);
+    ucs_assert(!(req->send.flush.started_lanes & UCS_BIT(lane)));
```
Comment (yosefe) [inline src/ucp/rma/flush.c:268]: maybe this assert not needed - already done in ucp_ep_flush_request_change_uct_cmp
or convert to assertv

File: src/ucp/wireup/wireup.c [hunk 29]
```diff
@@ -419,14 +419,125 @@ ucp_wireup_connect_local(ucp_ep_h ep,
     return status;
 }
 
-void ucp_wireup_remote_connect_lanes(ucp_ep_h ep, int ready)
+static int ucp_ep_has_wireup_pending(ucp_ep_h ucp_ep)
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:422]: ucp_ep_has_wireup_msg_pending

File: src/ucp/wireup/wireup.c [hunk 30]
```diff
@@ -419,14 +419,125 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
-        if (ucp_wireup_ep_test(ucp_ep_get_lane(ep, lane))) {
-            ucp_wireup_ep_remote_connected(ucp_ep_get_lane(ep, lane), ready);
+    for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
+        wireup_ep = ucp_wireup_ep(ucp_ep_get_lane(ucp_ep, lane));
+        if (wireup_ep == NULL) {
+            continue;
+        }
+
+        ucs_assert(wireup_ep->flags & UCP_WIREUP_EP_FLAG_READY);
+        ucs_assert(wireup_ep->super.uct_ep != NULL);
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:434]: move these to ucp_wireup_eps_progress

File: src/ucp/wireup/wireup.c [hunk 31]
```diff
@@ -419,14 +419,125 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+    ucp_wireup_ep_t *wireup_ep;
+    ucs_queue_head_t tmp_pending_queue;
+    uct_pending_req_t *uct_req;
+    ucp_lane_index_t lane;
+
+    UCS_ASYNC_BLOCK(&ucp_ep->worker->async);
+
+    /* If we still have pending wireup messages, send them out first */
+    if (ucp_ep_has_wireup_pending(ucp_ep)) {
+        ucp_wireup_eps_progress_sched(ucp_ep);
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:469]: /* Reschedule the progress since it's one-shot */

File: src/ucp/wireup/wireup.c [hunk 32]
```diff
@@ -419,14 +419,125 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+    }
+
+    return 0;
+}
+
+static void ucp_wireup_eps_progress_sched(ucp_ep_h ucp_ep)
+{
+    uct_worker_cb_id_t prog_id = UCS_CALLBACKQ_ID_NULL;
+
+    uct_worker_progress_register_safe(ucp_ep->worker->uct,
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:447]: /* Use one-shot mode to avoid the need to save prog_id */

File: src/ucp/wireup/wireup.c [hunk 33]
```diff
@@ -419,14 +419,125 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+    ucs_queue_head_init(&tmp_pending_queue);
+    for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
+        wireup_ep = ucp_wireup_ep(ucp_ep_get_lane(ucp_ep, lane));
+        if (wireup_ep == NULL) {
+            continue;
+        }
+
+        ucs_trace("ep %p: switching wireup_ep %p to ready state", ucp_ep,
+                  wireup_ep);
+
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:495]: remove space line

File: src/ucp/wireup/wireup.c [hunk 34]
```diff
@@ -419,14 +419,125 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+    for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
+        wireup_ep = ucp_wireup_ep(ucp_ep_get_lane(ucp_ep, lane));
+        if (wireup_ep == NULL) {
+            continue;
         }
+
+        ucs_assert(wireup_ep->super.uct_ep != NULL);
+        ucs_assert(wireup_ep->super.ucp_ep == ucp_ep);
+        ucs_assert(wireup_ep->flags & UCP_WIREUP_EP_FLAG_LOCAL_CONNECTED);
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:528]: see if some asserts are duplicate
the remaining ones to convert to assertv

File: src/ucp/wireup/wireup.c [hunk 35]
```diff
@@ -419,14 +419,125 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+        if (ready) {
+            wireup_ep->flags |= UCP_WIREUP_EP_FLAG_READY;
+        }
+
+        ucs_trace("ep %p: wireup ep %p is %s", ucp_ep, wireup_ep,
+                  ready ? "ready" : "remote-connected");
+    }
+
+    if (ready) {
+        ucp_wireup_eps_progress_sched(ucp_ep);
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:540]: can move this outside the function

File: src/ucp/wireup/wireup.c [hunk 36]
```diff
@@ -419,14 +419,125 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
         }
+
+        ucs_assert(wireup_ep->super.uct_ep != NULL);
+        ucs_assert(wireup_ep->super.ucp_ep == ucp_ep);
+        ucs_assert(wireup_ep->flags & UCP_WIREUP_EP_FLAG_LOCAL_CONNECTED);
+
+        wireup_ep->flags |= UCP_WIREUP_EP_FLAG_REMOTE_CONNECTED;
+        if (ready) {
+            wireup_ep->flags |= UCP_WIREUP_EP_FLAG_READY;
+        }
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:533]: maybe rename the func to ucp_wireup_update_flags(ucp_ep, flags)
that will add "flags" to all wireup_ep->flags

File: src/ucp/wireup/wireup.c [hunk 37]
```diff
@@ -419,14 +419,125 @@ ucp_wireup_connect_local(ucp_ep_h ep,
...
+            continue;
+        }
+
+        ucs_trace("ep %p: switching wireup_ep %p to ready state", ucp_ep,
+                  wireup_ep);
+
+        ucs_queue_for_each_extract(uct_req, &wireup_ep->pending_q, priv, 1) {
+            ucs_queue_push(&tmp_pending_queue, ucp_wireup_ep_req_priv(uct_req));
+            ucp_worker_flush_ops_count_dec(ucp_ep->worker);
+        }
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:499]: maybe add internal API to wireup_ep.h:
ucp_wireup_ep_pending_extract(ep, &pending_q)

File: src/ucp/core/ucp_worker.h [hunk 38]
```diff
@@ -402,21 +402,32 @@ ucs_status_t ucp_worker_discard_uct_ep_pending_cb(uct_pending_req_t *self);
...
 {
     UCP_WORKER_THREAD_CS_CHECK_IS_BLOCKED_CONDITIONAL(worker);
-    ucs_assert(worker->flush_ops_count < UINT_MAX);
-    ++worker->flush_ops_count;
+
+    if (count < 0) {
+        ucs_assert(((int64_t)worker->flush_ops_count + count) >= 0);
+    } else if (count > 0) {
+        ucs_assert(((uint64_t)worker->flush_ops_count + count) < UINT_MAX);
+    }
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.h:415]: long flush_ops_count = (long)worker->flush_ops_count + count;
ucs_assertv((flush_ops_count >= 0) && (flush_ops_count < INT_MAX), "worker->flush_ops_count=%d count=%d new_flush_ops_count=%ld", ...)

File: src/ucp/core/ucp_worker.h [hunk 39]
```diff
@@ -402,21 +402,32 @@ ucs_status_t ucp_worker_discard_uct_ep_pending_cb(uct_pending_req_t *self);
...
+
+    worker->flush_ops_count += count;
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_worker_flush_ops_count_inc(ucp_worker_h worker)
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.h:421]: can we remove the inc/dec wrappers now?

File: src/ucp/rma/flush.c [hunk 40]
```diff
@@ -16,25 +16,26 @@
 
 
 static void
-ucp_ep_flush_request_change_uct_cmp(ucp_request_t *req, int diff,
-                                    ucp_lane_map_t new_started_lanes)
+ucp_ep_flush_request_change_uct_comp(ucp_request_t *req, int diff,
```
Comment (yosefe) [inline src/ucp/rma/flush.c:19]: ucp_ep_flush_request_update_uct_comp

File: src/ucp/rma/flush.c [hunk 41]
```diff
@@ -16,25 +16,26 @@
...
 
     req->send.state.uct_comp.count += diff;
     req->send.flush.started_lanes  |= new_started_lanes;
 
     ucp_trace_req(req,
-                  "ep %p flush changed completion count by %d - result %d"
-                  "started lanes 0x%x by 0x%x", req->send.ep, diff,
-                  req->send.state.uct_comp.count, req->send.flush.started_lanes,
-                  new_started_lanes);
+                  "ep=%p flush changed completion count by %d - result %d, "
```
Comment (yosefe) [inline src/ucp/rma/flush.c:34]: "flush update ep %p comp_count %d->%d num_lanes %d->%d started_lanes 0x%x->0x%x"

File: src/ucp/rma/flush.c [hunk 42]
```diff
@@ -223,21 +216,22 @@ static unsigned ucp_ep_flush_resume_slow_path_callback(void *arg)
...
         /* Request was detached from pending and should be scheduled again */
         ucs_assertv(!(UCS_BIT(req->send.lane) & req->send.flush.started_lanes),
-                    "req %p: lane %d, started_lanes 0x%x", req,
-                    req->send.lane, req->send.flush.started_lanes);
+                    "req=%p lane=%d started_lanes=0x%x", req, req->send.lane,
+                    req->send.flush.started_lanes);
         /* Only lanes connected to iface can be started/flushed before
-         * wireup is done */
+         * wireup is done because connect2iface does not create wireup_ep
+         * without cm mode*/
```
Comment (yosefe) [inline src/ucp/rma/flush.c:228]: space after "mode"

File: src/ucp/wireup/wireup.c [hunk 43]
```diff
@@ -490,13 +486,13 @@ unsigned ucp_wireup_eps_progress(void *arg)
             continue;
         }
 
+        ucs_assert(wireup_ep->flags & UCP_WIREUP_EP_FLAG_READY);
+        ucs_assert(wireup_ep->super.uct_ep != NULL);
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c:490]: space line after

**Review by yosefe — DISMISSED** [review]
pls squash

**Comment by evgeny-leksikov** [comment]
> pls squash

done, rebased because of merge conflicts

**Review by yosefe — APPROVED** [review]
